### PR TITLE
Issue 1656 forms

### PR DESCRIPTION
**Pull request by alekpentchev**
# Description

Components - DynamicForm and Form together with their tests - were rewritten to support React Hooks API.

## Corresponding Issue

https://github.com/ifmeorg/ifme/issues/1656

# Screenshots

image

---

Reviewing this pull request? Check out our [Code Review Practices](https://github.com/ifmeorg/ifme/wiki/Code-Review-Practices) guide if you haven't already!

### CODE REVIEW

**Comment by alekpentchev** [comment]
One lint error exists because of eslint issue not resolved yet - https://github.com/yannickcr/eslint-plugin-react/issues/1280

This is also the reason why circleci fails.

**Comment by julianguyen** [comment]
Looks like there's legitimate rspec failures related to form submission! Let me know if you need any help debugging.

**Comment by julianguyen** [comment]
I did some refactors just now for duplicate code between Form and DynamicForm. Here's the merged [PR](https://github.com/ifmeorg/ifme/pull/1688). You'll need to rebase, let me know if you need any help!

**Comment by alekpentchev** [comment]
@julianguyen It's super strange - these failing tests occured after adding ```// eslint-disable-next-line ``` before extracting ```action``` from ```props```.
After removing this comment tests fail with the old error that ```action is missing in props validation```.

**Comment by alekpentchev** [comment]
For testing purposes I've added ```prop-types``` package to client's dependencies and defined propTypes for the Form component like this:
```
Form.propTypes = {
  action: PropTypes.string,
};
```
Eslint / Flow error about missing props in validation does not appear anymore. But I've mixed feelings about adding another dependency just to fix such strange and should-not-happen-at-all issue. What is your opinion on that topic?

**Comment by julianguyen** [comment]
I think we could away from Flow in the long-term since PropTypes from React offers a lot of what Flow does. You could try doing this for this PR and then we can do the same elsewhere in another PR.

**Comment by alekpentchev** [comment]
> I think we could move away from Flow in the long-term since PropTypes from React offers a lot of what Flow does. You could try doing this for this PR and then we can do the same elsewhere in another PR.

I see your point but I'm not sure whether moving away from Flow completely would be a good approach. In my opinion we'd lose important information about what types do state variables have, what type of data do functions return etc.

On the other hand I've asked a question on that matter (the issue with propTypes) directly in the eslint-react-plugin repo and I've received a such reply on the interactions between Flow types and propTypes:

> Both are required; neither flow nor typescript is capable of catching everything PropTypes can.

To sum up - I'll commit my changes containing added propTypes for the "problematic"component. We'll see then how circleci tests behave, too.

**Comment by julianguyen** [comment]
Yeah, the trade-off is losing state type information :/. But I think it should fine. We should rely on native React features as much as possible.

Sounds like a good plan!

**Comment by julianguyen** [comment]
I'm actually going to check out your branch and see if I can fix the linting issues without having to move away from Flow.

I think we can revisit the move to prop-types another time, outside of this pull request/task.

**Comment by alekpentchev** [comment]
@julianguyen Thank you very much for your support. I was leaning towards your direction and your implementation looks fine. I will review for myself to use it with the other components assigned to me.

+ I agree that we should discuss Flow in another thread.
